common/src/main/java/com/microsoft/hydralab/common/util/IOSPerfTestHelper.java

```diff
@@ -0,0 +1,117 @@
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class IOSPerfTestHelper {
+    private static IOSPerfTestHelper instance;
+
+    private Logger classLogger = LoggerFactory.getLogger(getClass());
+
+    private static final String lowestVersion = "0.10.2";
+    private static final String installCommand = "pip3 install -U tidevice[openssl]";
+    private Map<String, IOSPerfScriptProcessInfo> iOSPerfScriptProcesses;
+
+    private IOSPerfTestHelper() {
+        iOSPerfScriptProcesses = new ConcurrentHashMap<>();
+
+        // Check version of tidevice
+        boolean needUpdate = false;
+        String versionString = ShellUtils.execLocalCommandWithResult("tidevice -v", classLogger);
+        if (versionString == null) {
+            needUpdate = true;
+        } else {
+            String[] currentVersionParts = versionString.trim().split("\\.");
+            String[] lowestVersionParts = versionString.trim().split("\\.");
+            for (int i = 0; i < currentVersionParts.length; i++) {
+                int currentPart = Integer.parseInt(currentVersionParts[i]);
+                int requiredPart = Integer.parseInt(lowestVersionParts[i]);
+                if (currentPart < requiredPart) {
+                    needUpdate = true;
+                    break;
+                }
+            }
+        }
+        if (needUpdate) {
+            ShellUtils.execLocalCommand(installCommand, classLogger);
+        }
+    }
+
+    public void add(String key, File resultFile, Process process) {
+        if (!isRunning(key)) {
+            iOSPerfScriptProcesses.put(key, new IOSPerfScriptProcessInfo(resultFile, process));
+        }
+    }
+
+    public boolean isRunning(String key) {
+        return iOSPerfScriptProcesses.containsKey(key);
+    }
+
+    @Nullable
+    public File getResultFile(String key) {
+        IOSPerfScriptProcessInfo processInfo = iOSPerfScriptProcesses.get(key);
+        if (processInfo != null) {
+            File outputFile = processInfo.getResultFile();
+            return outputFile;
+        }
+        return null;
+    }
+
+    public long getStartTime(String key) {
+        IOSPerfScriptProcessInfo processInfo = iOSPerfScriptProcesses.get(key);
+        if (processInfo != null) {
+            return processInfo.getStartTimestamp();
+        }
+        return 0;
+    }
+
+    public void stop(String key) {
+        IOSPerfScriptProcessInfo processInfo = iOSPerfScriptProcesses.get(key);
+        if (processInfo != null) {
+            processInfo.stopProcess();
+            iOSPerfScriptProcesses.remove(key);
+        }
+    }
+
+    public static IOSPerfTestHelper getInstance() {
+        if (instance == null) {
+            instance = new IOSPerfTestHelper();
+        }
+        return instance;
```

Comment:
I think we can use double check lock to create the singleton. Especially in a multi-threaded environment.

Reply:
revised